### ISSUE TITLE
How to determine parameters for TF simulator?

### ISSUE THREAD

**Issue by krfricke** [park-project/park#6]
In [tf_sim.py](https://github.com/park-project/park/blob/master/park/envs/tf_placement_sim/tf_sim.py#L53) hardcoded default parameters are given for the AWS p2.8xlarge instance. I'd like to use the simulator for other instances (i.e. google cloud machines) and was wondering how these parameters were determined. Were they measured from the real hardware? Is there a script or code for that?

**Comment by hongzimao** [comment]
Thanks for your interest! Those parameters for the runtime were measured from real hardware. I think @aravic has some scripts for collecting those numbers. He will help you if that part of the process is fully automated.

**Comment by aravic** [comment]
Hi,

Thanks for your interest.

A simple grid search was used to find the parameters to minimize a loss function (pasted down below).
I think that if you've a few tensorflow graphs and their corresponding step stats collected, you should be able to get these numbers by a similar grid search for a different device configuration. Alternatively, some of the main parameters here (say transfer_speed) can be directly extracted from the hardware specs of target devices.


`
  def get_loss(self, node_dict, step_stats, verbose=False):
    """
      Given the node_dict and step_stats, computes loss
      Loss is the L2 norm of the difference in actual and
      simulated start and end times

    node_dict: mapping from node name to its compute cost.
    """
    stats = defaultdict(self.Node)
    min_time = util.get_min_time(step_stats, lambda x: "all" in x or "memcpy" in x)

    # Compress nodes in step_stats
    for dev_stats in step_stats.dev_stats:
      if "all" in dev_stats.device or "CPU" in dev_stats.device:
        for s in dev_stats.node_stats:
          node_name = s.node_name.split(":")[0]
          s_start = s.all_start_micros - min_time
          s_end = s_start + s.op_end_rel_micros
          if node_name in stats:
            s_start = min(s_start, stats[node_name].start_time)
            s_end = max(s_end, stats[node_name].end_time)
          stats[node_name].start_time = s_start
          stats[node_name].end_time = s_end
          stats[node_name].compute_cost = s_end - s_start

    info = self.Node()
    info.node_count = 0
    info.avg_diff = 0
    info.loss = 0
    for node_name, s_node in sorted(stats.items(), key=lambda x : x[1].start_time):
      if node_name not in node_dict:
        continue
      n_node = node_dict[node_name]
      info.node_count += 1
      info.loss += (s_node.start_time - n_node.start_time)**2
      info.loss += (s_node.end_time - n_node.end_time)**2
      info.avg_diff += s_node.start_time - n_node.start_time
      if verbose:
        print(
            "\t", s_node.start_time,
            # "\t", s_node.end_time, 
            "\t", int(n_node.start_time),
            # "\t", int(n_node.end_time),
            "\t", s_node.start_time - int(n_node.start_time),
            # "\t", s_node.end_time - int(n_node.end_time),
            "\t", n_node.device,
            "\t", s_node.compute_cost,
            "\t", int(n_node.compute_cost),
            # "\t", n_node.output_memory,
            "\t", node_name
        )
    info.loss = math.sqrt(info.loss / info.node_count / 2)
    info.avg_diff = info.avg_diff / info.node_count
    return info`

**Comment by krfricke** [comment]
Thank you very much for your help. I'll try to apply this to our setup later this week or next weekend and will report back on how it went.

**Comment by krfricke** [comment]
Hi, I finally got around to trying to estimate the parameters.

I used this script to collect the run metadata/step stats:

```python
import time

import tensorflow as tf

from tensorflow.core.protobuf import rewriter_config_pb2
from tensorflow.python.framework import meta_graph, ops as tf_ops


def profile_graph(graph, call_op=None, feed_dict=None, init_fn=None):
    """
    Profile tensorflow graph, returns run metadata (including step stats) and metagraph
    Args:
        graph: TensorFlow graph object
        call_op: Op to call. If None, tries to uses ops of the TRAIN_OP collection
        feed_dict: Feed dict to pass to session call
        init_fn: Optional init fn

    Returns: Tuple of (run_metadata, metagraph)

    """
    session_config = tf.compat.v1.ConfigProto(
        allow_soft_placement=True,
        log_device_placement=False,
        inter_op_parallelism_threads=0,
        intra_op_parallelism_threads=0,
        isolate_session_state=True
    )
    session_config.graph_options.rewrite_options.arithmetic_optimization = rewriter_config_pb2.RewriterConfig.OFF
    session_config.graph_options.rewrite_options.memory_optimization = rewriter_config_pb2.RewriterConfig.OFF
    session_config.gpu_options.allow_growth = True

    sess = tf.Session(target="", config=session_config, graph=graph)

    sess.run(tf.global_variables_initializer())
    sess.run(tf.tables_initializer())
    if init_fn:
        init_fn(sess)

    if not call_op:
        call_op = graph.get_collection_ref(tf_ops.GraphKeys.TRAIN_OP)

    run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
    for i in range(3):  # at least 1 warmup
        run_metadata = tf.RunMetadata()  # Always use a fresh RunMetadata object
        start_time = time.time()
        step_result = sess.run(
            [call_op],
            feed_dict=feed_dict,
            options=run_options,
            run_metadata=run_metadata
        )
        time_taken = time.time() - start_time
        print("Took {:.2f} seconds to call session".format(time_taken))

    metagraph = meta_graph.create_meta_graph_def(graph=graph)

    return run_metadata, metagraph
```

and then altered the script you provided to do a simple grid search:

```python
import copy
import itertools
import math

import numpy as np

from collections import defaultdict
from tqdm import tqdm

from park.envs.tf_placement_sim.tf_sim import Simulator, get_op_costs


def grid_search(**kwargs):
    """
    Iterate through parameter values in an exhaustive grid-search.

    Args:
        **kwargs: dict in the form { parameter_name: iterator, ... }

    Returns:
        Iterator of dicts in the form { parameter_name: value, ... }

    """
    parameter_names = kwargs.keys()
    for parameter_values in itertools.product(*kwargs.values()):
        yield dict(zip(parameter_names, parameter_values))


class ParameterEstimator(object):
    def __init__(self, devices, graphs, exclude_ops=None):
        """
        Args:
            graphs: list of tuples (metagraph, step_stats)
        """
        self.devices = devices
        self.graphs = graphs
        self.exclude_ops = exclude_ops

    def estimate(self):
        for i, (metagraph_orig, step_stats) in enumerate(self.graphs):
            print("Searching parameters for graph {} of {}:".format(i+1, len(self.graphs)))
            metagraph = copy.deepcopy(metagraph_orig)

            delete_ids = []
            for i, node in enumerate(metagraph.graph_def.node):
                if node.name in self.exclude_ops:
                    print("Deleting node {}: {}".format(i, node.name))
                    delete_ids.append(i)

            for i in sorted(delete_ids, reverse=True):
                del metagraph.graph_def.node[i]

            cost_dict, _ = get_op_costs(step_stats)

            min_loss = -1.0
            min_params = None
            for params in tqdm(grid_search(
                    delta1=np.arange(18, 25, 1)/10.,  # Avoid floating point precision errors
                    delta2=range(60, 150, 1)
            ), total=7*90):  # Calculate total manually for nice tqdm output
                sim = Simulator(metagraph, cost_dict, {}, self.devices)
                sim.params.update(params)

                device_dict = {}
                for node_name, node in sim.node_dict.items():
                    if not hasattr(node, 'device') or node.device == '':
                        device_dict[node_name] = self.devices[0]
                        # node.device = self.devices[0]
                    else:
                        device_dict[node_name] = node.device

                run_time, node_dict = sim.simulate(device_dict)

                loss_node = self.get_loss(node_dict, step_stats)

                # print("Testing params: {}, found loss of {:.4f}".format(str(params), loss_node.loss))

                if loss_node.loss < min_loss or min_loss == -1:
                    min_loss = loss_node.loss
                    min_params = params

            print("Found smallest loss of {:.4f} with params {}".format(min_loss, min_params))

    class Node(object):
        pass

    def get_min_time(self, step_stats, filter_callback):
        """

        Args:
            step_stats:
            filter_callback: Filter device names by this callback

        Returns:

        """
        min_time = -1
        for dev_stats in step_stats.dev_stats:
            # Filter by callback
            if filter_callback and not filter_callback(dev_stats.device):
                continue
            # Collect minimum time
            for node_stats in dev_stats.node_stats:
                if node_stats.all_start_micros < min_time or min_time == -1:
                    min_time = node_stats.all_start_micros

        return min_time


    def get_loss(self, node_dict, step_stats, verbose=False):
        """
          Given the node_dict and step_stats, computes loss
          Loss is the L2 norm of the difference in actual and
          simulated start and end times

        Args:
            node_dict: mapping from node name to its compute cost.
            step_stats: Step stats for a tf graph in tensorflow step stats format (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/step_stats.proto).
            verbose:

        Returns:
            info Node object containing loss, average difference, and node count
        """


        stats = defaultdict(self.Node)
        min_time = self.get_min_time(step_stats, lambda x: "all" in x or "memcpy" in x)

        # Compress nodes in step_stats
        for dev_stats in step_stats.dev_stats:
            if "all" in dev_stats.device or "CPU" in dev_stats.device:
                for s in dev_stats.node_stats:
                    node_name = s.node_name.split(":")[0]
                    s_start = s.all_start_micros - min_time
                    s_end = s_start + s.op_end_rel_micros
                    if node_name in stats:
                        s_start = min(s_start, stats[node_name].start_time)
                        s_end = max(s_end, stats[node_name].end_time)
                    stats[node_name].start_time = s_start
                    stats[node_name].end_time = s_end
                    stats[node_name].compute_cost = s_end - s_start

        info = self.Node()
        info.node_count = 0
        info.avg_diff = 0
        info.loss = 0
        for node_name, s_node in sorted(stats.items(), key=lambda x: x[1].start_time):
            if node_name not in node_dict:
                continue
            n_node = node_dict[node_name]

            # Skip nodes that have not been evaluated
            if not hasattr(n_node, 'start_time'):
                continue

            info.node_count += 1

            info.loss += (s_node.start_time - n_node.start_time) ** 2
            info.loss += (s_node.end_time - n_node.end_time) ** 2
            info.avg_diff += s_node.start_time - n_node.start_time
            if verbose:
                print(
                    "\t", s_node.start_time,
                    # "\t", s_node.end_time,
                    "\t", int(n_node.start_time),
                    # "\t", int(n_node.end_time),
                    "\t", s_node.start_time - int(n_node.start_time),
                    # "\t", s_node.end_time - int(n_node.end_time),
                    "\t", n_node.device,
                    "\t", s_node.compute_cost,
                    "\t", int(n_node.compute_cost),
                    # "\t", n_node.output_memory,
                    "\t", node_name
                )
        info.loss = math.sqrt(info.loss / info.node_count / 2)
        info.avg_diff = info.avg_diff / info.node_count
        return info
```

I altered your loss code only slightly, specifically there were some nodes that were not processed in the simulation (while loops) and I just made sure to exclude them.

The script works, although my grid search returns rather high delay values. Do you have any step stats measurements available from your hardware measurements so I could try to see if I can replicate the parameter values? I found https://github.com/park-project/park/blob/master/park/envs/tf_placement_sim/graphs/inception.pkl which has a step_stats attribute, has that been measured on one of the AWS instances?

Also, I didn't find where I can obtain the transfer_speed. Can you point me to the right direction? We use K80 GPUs like AWS does, so we might not need to adjust this parameter after all.

Again, thanks for your help!

**Comment by krfricke** [comment]
Quick additional question: Do I have to collect the step stats for each node on each device separately? I.e. first schedule all nodes on CPU:0, make a measurement, then on GPU:0 etc. 

In the code it seems like the execution stats are collected for just one device, but it would make sense to me to take different device execution times into account.

**Comment by aravic** [comment]
Hi, Sorry for a delayed response. I am not an active member of this project anymore but I would be happy to answer your questions.

The step_stats found under `tf_placement_sim/graphs/inception.pkl` has indeed been measured on p2.8xlarge instance. 

Wouldn't it be possible to tune the `transfer_speed` parameter by grid search over a reasonable search space? 

We always used homogeneous device configs and so measuring on one of them was sufficient. We did not consider placements on CPU device.
